modules/core/lib/Auth/Process/PairwiseID.php

```diff
@@ -0,0 +1,134 @@
+ * authproc-configuration.
+ *
+ * Example - generate from attribute:
+ * <code>
+ * 'authproc' => [
+ *   50 => [
+ *       'core:PairwiseID',
+ *       'identifyingAttribute' => 'uid',
+ *       'scope' => 'example.org',
+ *   ]
+ * ]
+ * </code>
+ *
+ * @package SimpleSAMLphp
+ */
+class PairwiseID extends Auth\ProcessingFilter
+{
+    /**
+     * The attribute we should generate the pairwise id from.
+     *
+     * @var string
+     */
+    private string $identifyingAttribute;
+
+    /**
+     * The scope to use for this attribute.
+     *
+     * @var string
+     */
+    private string $scope;
+
+    /**
+     * @var \SimpleSAML\Utils\Config|string
+     * @psalm-var \SimpleSAML\Utils\Config|class-string
+     */
+    protected $configUtils = Utils\Config::class;
+
+
+    /**
+     * Initialize this filter.
+     *
+     * @param array &$config  Configuration information about this filter.
+     * @param mixed $reserved  For future use.
+     */
+    public function __construct(array &$config, $reserved)
+    {
+        parent::__construct($config, $reserved);
+
+        Assert::keyExists($config, 'identifyingAttribute', "Missing mandatory 'identifyingAttribute' config setting.");
+        Assert::keyExists($config, 'scope', "Missing mandatory 'scope' config setting.");
+        Assert::stringNotEmpty($config['identifyingAttribute']);
+        Assert::regex(
+            $config['scope'],
+            '/^[a-zA-Z0-9.-]+$/i',
+            'PairwiseID: \'scope\' contains illegal characters.'
+        );
+
+        $this->identifyingAttribute = $config['identifyingAttribute'];
+        $this->scope = $config['scope'];
+    }
+
+
+    /**
+     * Apply filter to add the Pairwise ID.
+     *
+     * @param array &$state  The current state.
+     */
+    public function process(array &$state): void
+    {
+        Assert::keyExists($state, 'Attributes');
+        Assert::keyExists(
+            $state['Attributes'],
+            $this->identifyingAttribute,
+            sprintf(
+                "core:PairwiseID: Missing attribute '%s', which is needed to generate the Pairwise ID.",
+                $this->identifyingAttribute
+            )
+        );
+
+        $userID = $state['Attributes'][$this->identifyingAttribute][0];
+        Assert::regex(
+            $userID,
+            '/^[a-zA-Z0-9=-]+$/i',
+            'PairwiseID: \'identifyingAttribute\' contains illegal characters.'
+        );
+
+        $idp_entityid = $state['IdPMetadata']['entityid'];
+
+        if (!empty($state['saml:RequesterID'])) {
+            // Proxied request - use actual SP entity ID
+            $sp_entityid = $state['saml:RequesterID'][0];
+        } else {
+            $sp_entityid = $state['core:SP'];
+        }
+
+        // Calculate hash
+        $salt = $this->configUtils::getSecretSalt();
+        $hash = hash('sha256', $salt . '|' . $userID . '|' . $idp_entityid . '|' . $sp_entityid, false);
```

Comment:
I wonder if we should add the idp_entityid into the hash. This obviously means any identifier will change when the idp entityid changes.

I think we can and/or should assume (and document) that the identifying attribute must be unique for the given scope (not per se for the given idp). So that keeping the scope the same will keep the identifiers the same.

Reply:
I've discussed adding the idp_entity with @jaimeperez and he insisted to replicate the original EPTiD-behaviour.. I agree with you though.

Reply:
Maybe I'm missing the rationale for that then. Because I don't understand it currently. For SubjectID, we find it acceptable to conclude that identifyingAttribute + scope is globally unique. For pairwiseID, we insist to add idp entityid to the mix. I don't understand why we make this difference.

Reply:
Jaime's argument was that it's a tie between IDP and SP... I agree we can totally do without.

Reply:
I do not see a base for that in the spec.

Reply:
I'll poke him once more on this.. See we we can settle this

Reply:
Removed the IDP entityid from the hash